src/op_analytics/coreutils/logger.py

```diff
@@ -109,3 +109,37 @@ def human_interval(num_seconds: int) -> str:
 def memory_usage():
     """Return max_rss / 1e6 rounded to make it easier to eyeball."""
     return round(resource.getrusage(resource.RUSAGE_SELF).ru_maxrss / 1e6, 2)
+
+
+class ProgressTracker:
+    """
+    Tracks progress and ETA for a batch of tasks.
+    Instead of logging a separate "Task progress" line, we bind context variables
+    so that when the wrapped function logs, the ETA and progress fields are included
+    in that log line.
+    """
+
+    def __init__(self, total_tasks: int):
+        self.total_tasks = total_tasks
+        self.completed_tasks = 0
+        self.start_time = now()
+
+    def wrap(self, func, current_index: int):
+        def wrapper(target_item):
+            self.completed_tasks += 1
```

Comment:
the estimate should be done after the function has returned, and maybe only reported if you have already completed more than 3 tasks.